Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -484,140 +483,205 @@ void TextureCacheBase::UnbindTextures()
-  const unsigned int bsh = TexDecoder_GetBlockHeightInTexels(texformat);
-
-  unsigned int expandedWidth = Common::AlignUp(width, bsw);
-  unsigned int expandedHeight = Common::AlignUp(height, bsh);
-  const unsigned int nativeW = width;
-  const unsigned int nativeH = height;
-
-  // Hash assigned to texcache entry (also used to generate filenames used for texture dumping and
-  // custom texture lookup)
-  u64 base_hash = TEXHASH_INVALID;
-  u64 full_hash = TEXHASH_INVALID;
-
-  u32 full_format = texformat;
+  struct RequestedTextureLevel
+  {
+    u32 width;
+    u32 height;
+    u32 padded_width;
+    u32 padded_height;
+    const u8* data;
+    const u8* data_gb;  // red/green data for rgba8_from_tmem textures
+    u32 total_size;
+  };
+
+  static TextureRequest FromBPMem(const u32 stage)
+  {
+    const FourTexUnits& tex = bpmem.tex[stage >> 2];
+    const u32 id = stage & 3;
 
-  const bool isPaletteTexture =
-      (texformat == GX_TF_C4 || texformat == GX_TF_C8 || texformat == GX_TF_C14X2);
+    return TextureRequest(tex.texMode0[id], tex.texMode1[id], tex.texImage0[id], tex.texImage1[id],
+                          tex.texImage2[id], tex.texImage3[id], tex.texTlut[id]);
+  }
 
-  // Reject invalid tlut format.
-  if (isPaletteTexture && tlutfmt > GX_TL_RGB5A3)
-    return nullptr;
+  TextureRequest(TexMode0 mode0, TexMode1 mode1, TexImage0 image0, TexImage1 image1,
+                 TexImage2 image2, TexImage3 image3, TexTLUT tlut)
+  {
+    m_from_tmem = image1.image_type != 0;
+    m_format = image0.format;
+    m_has_tlut = m_format == GX_TF_C4 || m_format == GX_TF_C8 || m_format == GX_TF_C14X2;
+    m_tlut = m_has_tlut ? &texMem[tlut.tmem_offset << 9] : nullptr;
+    m_tlut_size = TexDecoder_GetPaletteSize(image0.format);
+    m_tlut_format = static_cast<TlutFormat>(m_has_tlut ? tlut.tlut_format : 0);
+
+    // "m_address" is provided, even if loading from tmem, as it is used as a cache key.
+    // I'm not sure if that's sensible, but it's necessary for some FIFO logs.
+    m_address = image3.image_base << 5;
+    m_rbga8_from_tmem = m_from_tmem && m_format == GX_TF_RGBA8;
+    m_use_mipmaps = SamplerCommon::AreBpTexMode0MipmapsEnabled(mode0);
+
+    const u32 width = image0.width + 1;
+    const u32 height = image0.height + 1;
+
+    if (m_use_mipmaps)
+    {
+      u32 levels = ((mode1.max_lod + 0xF) / 0x10 + 1);
+      // GPUs don't like when the specified mipmap count would require more than one 1x1-sized LOD
+      // in the mipmap chain
+      // e.g. 64x64 with 7 LODs would have the mipmap chain 64x64,32x32,16x16,8x8,4x4,2x2,1x1,0x0,
+      // so we limit the mipmap count to 6 there
+      m_num_levels = std::min<u32>(IntLog2(std::max(width, height)) + 1, levels);
+    }
+    else
+    {
+      m_num_levels = 1;
+    }
 
-  if (isPaletteTexture)
-    full_format = texformat | (tlutfmt << 16);
+    const u8* src_data = m_from_tmem ? nullptr : Memory::GetPointer(m_address);
```

Comment:
Off topic - I'll admit, I don't usually const'ify most of my variables unless it's a pass in from a function.  What's the gain?
More on topic - maybe this optimization can already be done by the compiler.  But It seems odd we're branching 3 times to define some variables.  I almost wonder if the for loop could go into separate function called InitializeLevels().  Then you could do:

```
if (m_from_tmem)
{
  InitializeLevels(nullptr, &texMem[image1.tmem_even * TMEM_LINE_SIZE]), &texMem[image2.tmem_odd * TMEM_LINE_SIZE]);
}
else
{
  InitializeLevels(Memory::GetPointer(m_address), nullptr, nullptr);
}
```

Or potentially to mimic your approach with DecodeLevel, we have InitializeLevel and keep the for loop?

I don't feel very strongly about this though, just looking to improve code clarity and performance.